[RFC] Fix location list updates.

llorens opened the pull request:
[reproduce.txt](https://github.com/neovim/neovim/files/1347011/reproduce.txt)
Fix quickfix performance optimization which prevented quickfix items
from being updated when there were multiple windows with location lists
but the buffer with errors only in one of the lists.

Attached a script reproducing the issue.

KillTheMule commented:
Seems like one of the tests isn't happy. I'll leave the error message, since travis' logs are kinda annoying to search through:

```
From test_quickfix.vim:
Found errors in Test_adjust_lnum():
function RunTheTest[13]..Test_adjust_lnum[2]..Xadjust_qflnum line 21: Expected 13 but got 11
```

Could your script be made into a test? If you need help with that, let me know :)

justinmk commented:
Thanks, are you sure this isn't covered by an upstream vim patch? Would be better to merge those instead. They will also contain test coverage.

llorens committed:
quickfix: fix location list updates.

Fix quickfix performance optimization which prevented quickfix items
from being updated when there were multiple windows with location lists
but the buffer with errors only in one of the lists.

llorens commented:
Somehow I managed to fix one problem by introducing another one. Yes, I did not run the tests - sorry.
Regarding a fix upstream: there is none so far.

llorens commented:
@KillTheMule I can work on converting it into a test but I have very little time and I don't even know where to put it (would test/functional/ui/quickfix_spec.lua be a good place?). I assume src/nvim/testdir/test_quickfix.vim is not the right place since these are Vim's tests and Neovim has its own test system.

KillTheMule commented:
`test/functional/ui/quickfix_spec.lua ` sounds quite right ;)

chrisbra commented:
> Regarding a fix upstream: there is none so far.

Then perhaps you should raise an issue there?

llorens committed:
quickfix: fix location list updates (test).

llorens commented:
@KillTheMule I figured the test has nothing to do with the UI so I put it in the core directory (quickfix is core functionality, right?). Yes, I am low on time, but this was a good chance to look at the testing framework in Neovim.

KillTheMule commented on test/functional/core/quickfix_spec.lua:
```diff
@@ -0,0 +1,31 @@
+	call setloclist(0, [qf_item])
+	1del _
+	call append(0, ['New line 1', 'New line 2', 'New line 3'])
+	silent ll
```
This should be indented for better readability.

llorens replied:
Yes, these were tabs. Sorry.

KillTheMule commented:
I don't really mind, but since there already was a file for quickfix testing, that seemed the canonical place to put it. The testing framework is awesome alright :)

justinmk commented:
@llorens If I revert the changes to mark.c the test doesn't fail.

llorens commented:
@justinmk You have to revert the changes in quickfix.c as well. In fact I pushed the change as 2 separate commits: 1st one introduces a failing test and the 2nd one fixes the problem and makes the test succeed. Can you verify by simply checking out the 1st of the commits?
